### ISSUE TITLE
Question about the saved models

### ISSUE THREAD

**Issue by kunjiefan** [KarchinLab/bigmhc#2]
Hi,

There are several saved model checkpoints under model/ directory. Are these models trained on training data only (el_train.csv), or trained on both training and test data (el_train.csv and el.csv).

I ran your code to re-train the model using the training data for 50 epochs and evaluated on the test data. However, I could not obtain the same results as directly running the saved models you provided. I could only achieve AUPR of ~0.75.

**Comment by benjaminalbert** [comment]
Hi @kunjiefan,

Have you trained using the latest code and data?

If not, please pull the latest code and get the updated datasets as we have made many fixes and improvements described in our latest preprint revision: https://doi.org/10.1101/2022.08.29.505690

**Comment by kunjiefan** [comment]
Hi Benjamin,

Thanks for your quick reply. I did use the latest code and data.

When I directly ran the saved model bat8192/ on the test set (el.csv), I would get a AUROC of 0.994 and AUPRC of 0.957, which is higher than your reported results in the manuscript.

**Comment by benjaminalbert** [comment]
Hi @kunjiefan,

The models in this GitHub repository are the final EL models trained using the revised procedure, illustrated in Fig 1 of the revised preprint and detailed on pages 17-19. To evaluate on EL test data (`el_test.csv`), we create models using only the EL training (`el_train.csv`) and EL validation (`el_val.csv`) data. However, after evaluating BigMHC on EL testing data, we train finalized EL models on all EL data, which is why you see increased AUROC and AUPRC.

For reproducibility, we have uploaded the EL models that we trained using only `el_train.csv` and `el_val.csv` to the [Mendeley data repository](doi.org/10.17632/dvmz6pkzvb.3) in the file called `eltrainval_models.zip`. We have also uploaded the BigMHC predictions of these models, along with the predictions of prior methods, as columns in the `el_test.csv` file; these data are what we present in the preprint for EL results.

I have just reconfirmed that one can reproduce these EL results in the following manner. First, download and extract the `eltrainval_models.zip` and the `datasets.zip` from the Mendeley repository. Then, you can make predictions with the downloaded models as follows:

`python predict.py -i=el_test.csv -m=${modelsdir}/bat512/:/${modelsdir}/bat1024/:${modelsdir}/bat2048/:${modelsdir}/bat4096/:${modelsdir}/bat8192/:${modelsdir}/bat16384/:${modelsdir}/bat32768/`

Where I set `modelsdir` to the path to the extracted `eltrainval_models` dir.

Running the models from `eltrainval_models` produces an output column called `BigMHC`; compare this column against the `BigMHC_EL` column that we provide in `el_test.csv`.

I have just tested this again and I get a maximum absolute difference between the two columns of 5e-9 (due to floating point error).

Please let me know if this answers your questions.

**Comment by kunjiefan** [comment]
Hi Benjamin,

Thanks for your detailed guidance. I just realized that I'm using the old data (version 1), though I don't think the training data and test set are much different from the latest version. 

BTW, the data link in the README.md is still for version 1. You may need to update the link to version 3.

I'm curious about how much improvement the ensemble of seven models could achieve over a single model trained using a fixed batch size. Do you have some results?

Thanks again for your help.

**Comment by benjaminalbert** [comment]
Thanks for catching the README link; it now points to the latest Mendeley version.

There were many other improvements made to BigMHC in addition to the ensembling. Among them include 9:1 train:val splits to choose epochs that maximize AUPRC on validation data, slightly larger pseudosequences, and removal of the "Narrow LSTM" (i.e. canonical LSTM implementation, discussed in the revised manuscript ablation study).

You can test the results of individual models from the ensemble similar to the instructions listed above, but just specifying a single model like this:

`python predict.py -i=el_test.csv -m=${modelsdir}/bat512/`

Also, for immunogenicity prediction, we optimized batch size in addition to training epochs to maximize AUPRC on validation data. The results of validation can be found here: https://github.com/KarchinLab/bigmhc/tree/master/data/val

I will mark this issue as resolved, but feel free to open more issues and/or contact for further guidance.
